File: web/src/pages/canvas/vertical.ts
/**
 * Vertical (竖排) typesetting — mirror of `server/mio_server/pipeline/vertical.py`.
 * Keep the tables in sync; both test suites pin the same samples.
 */

export const VERTICAL_FORMS: Record<string, string> = {
  '，': '︐',
  ',': '︐',
  '、': '︑',
  '。': '︒',
  '：': '︓',
  ':': '︓',
  '；': '︔',
  ';': '︔',
  '！': '︕',
  '!': '︕',
  '？': '︖',
  '?': '︖',
  '…': '︙',
  '‥': '︰',
  '—': '︱',
  '–': '︲',
  '（': '︵',
  '(': '︵',
  '）': '︶',
  ')': '︶',
  '｛': '︷',
  '{': '︷',
  '｝': '︸',
  '}': '︸',
  '〔': '︹',
  '〕': '︺',
  '【': '︻',
  '】': '︼',
  '《': '︽',
  '》': '︾',
  '〈': '︿',
  '〉': '﹀',
  '「': '﹁',
  '」': '﹂',
  '『': '﹃',
  '』': '﹄',
  '［': '﹇',
  '[': '﹇',
  '］': '﹈',
  ']': '﹈',
  ー: '丨',
  '～': '≀',
  '~': '≀',
};

const COMBINED: Record<string, string> = {
  '!!': '‼',
  '！！': '‼',
  '!?': '⁉',
  '！？': '⁉',
  '?!': '⁈',
  '？！': '⁈',
  '??': '⁇',
};
const NO_START = new Set('︐︑︒︓︔︕︖︙︰︶︸︺︼︾﹀﹂﹄﹈‼⁉⁈⁇');
const NO_END = new Set('︵︷︹︻︽︿﹁﹃﹇');

/** Glyphs in vertical form; whitespace dropped, `\n` kept as a break. */
export function glyphs(text: string): string[] {
  const out: string[] = [];
  const chars = Array.from(text);
  let i = 0;
  while (i < chars.length) {
    const pair = chars[i] + (chars[i + 1] ?? '');
    if (COMBINED[pair]) {
      out.push(COMBINED[pair]);
      i += 2;
      continue;
    }
    const ch = chars[i];
    i += 1;
    if (ch === '\n') out.push('\n');
    else if (/\s/.test(ch)) continue;
    else out.push(VERTICAL_FORMS[ch] ?? ch);
  }
  return out;
}

/** Balanced columns of at most `maxChars` cells; the first column is the rightmost. */
export function columns(text: string, maxChars: number): string[][] {
  const limit = Math.max(2, maxChars);
  const cols: string[][] = [];
  const paras: string[][] = [[]];
  for (const g of glyphs(text)) {
    if (g === '\n') paras.push([]);
    else paras[paras.length - 1].push(g);
  }
  for (const cells of paras) {
    if (!cells.length) continue;
    const n = Math.ceil(cells.length / limit);
    const size = Math.ceil(cells.length / n);
    const part: string[][] = [];
    for (let k = 0; k < cells.length; k += size) part.push(cells.slice(k, k + size));
    for (let k = 1; k < part.length; k++) {
      while (part[k].length && NO_START.has(part[k][0])) part[k - 1].push(part[k].shift()!);
      while (part[k - 1].length > 1 && NO_END.has(part[k - 1][part[k - 1].length - 1]))
        part[k].unshift(part[k - 1].pop()!);
    }
    cols.push(...part.filter((p) => p.length));
  }
  return cols;
}

export function blockSize(cols: string[][], fontSize: number, colGap: number): [number, number] {
  if (!cols.length) return [0, 0];
  const w = cols.length * fontSize + (cols.length - 1) * colGap;
  const h = Math.max(...cols.map((c) => c.length)) * Math.round(fontSize * 1.05);
  return [w, h];
}

/** Cell centres `(glyph, x, y)`: columns right → left, glyphs top → bottom. */
export function cellCenters(
  cols: string[][],
  center: [number, number],
  fontSize: number,
  colGap: number,
): { ch: string; x: number; y: number }[] {
  const [w, h] = blockSize(cols, fontSize, colGap);
  const step = Math.round(fontSize * 1.05);
  const out: { ch: string; x: number; y: number }[] = [];
  cols.forEach((col, k) => {
    const x = center[0] + w / 2 - fontSize / 2 - k * (fontSize + colGap);
    col.forEach((ch, j) => out.push({ ch, x, y: center[1] - h / 2 + step * j + step / 2 }));
  });
  return out;
}

File: server/mio_server/pipeline/vertical.py
"""Vertical (竖排) typesetting: presentation forms, kinsoku and right-to-left columns.

Mirrored by ``web/src/pages/canvas/vertical.ts`` so the canvas preview matches the export; keep
the two tables in sync (both have tests that pin the same samples).
"""

from __future__ import annotations

import math

# Horizontal punctuation → vertical presentation forms (U+FE10–FE19, U+FE30–FE4F).
VERTICAL_FORMS = {
    "，": "︐",
    ",": "︐",
    "、": "︑",
    "。": "︒",
    "：": "︓",
    ":": "︓",
    "；": "︔",
    ";": "︔",
    "！": "︕",
    "!": "︕",
    "？": "︖",
    "?": "︖",
    "…": "︙",
    "‥": "︰",
    "—": "︱",
    "–": "︲",
    "（": "︵",
    "(": "︵",
    "）": "︶",
    ")": "︶",
    "｛": "︷",
    "{": "︷",
    "｝": "︸",
    "}": "︸",
    "〔": "︹",
    "〕": "︺",
    "【": "︻",
    "】": "︼",
    "《": "︽",
    "》": "︾",
    "〈": "︿",
    "〉": "﹀",
    "「": "﹁",
    "」": "﹂",
    "『": "﹃",
    "』": "﹄",
    "［": "﹇",
    "[": "﹇",
    "］": "﹈",
    "]": "﹈",
    "ー": "丨",
    "～": "≀",
    "~": "≀",
}
# Pairs set upright in a single cell (縦中横).
COMBINED = {"!!": "‼", "！！": "‼", "!?": "⁉", "！？": "⁉", "?!": "⁈", "？！": "⁈", "??": "⁇"}
# Glyphs that must not start a column (closing marks hang at the end of the previous one).
NO_START = set("︐︑︒︓︔︕︖︙︰︶︸︺︼︾﹀﹂﹄﹈‼⁉⁈⁇")
# Glyphs that must not end a column (opening marks move down to the next one).
NO_END = set("︵︷︹︻︽︿﹁﹃﹇")


def glyphs(text: str) -> list[str]:
    """Characters of ``text`` in vertical form, whitespace dropped, ``\\n`` kept as a break."""
    out: list[str] = []
    i = 0
    while i < len(text):
        pair = text[i : i + 2]
        if pair in COMBINED:
            out.append(COMBINED[pair])
            i += 2
            continue
        ch = text[i]
        i += 1
        if ch == "\n":
            out.append("\n")
        elif ch.isspace():
            continue
        else:
            out.append(VERTICAL_FORMS.get(ch, ch))
    return out


def columns(text: str, max_chars: int) -> list[str]:
    """Split into balanced columns of at most ``max_chars`` cells (first column = rightmost)."""
    max_chars = max(2, max_chars)
    cols: list[str] = []
    for para in "".join(glyphs(text)).split("\n"):
        cells = list(para)
        if not cells:
            continue
        n = math.ceil(len(cells) / max_chars)
        size = math.ceil(len(cells) / n)
        part: list[list[str]] = [cells[k : k + size] for k in range(0, len(cells), size)]
        for k in range(1, len(part)):
            while part[k] and part[k][0] in NO_START:  # hang closing marks
                part[k - 1].append(part[k].pop(0))
            while len(part[k - 1]) > 1 and part[k - 1][-1] in NO_END:
                part[k].insert(0, part[k - 1].pop())
        cols.extend("".join(p) for p in part if p)
    return cols


def block_size(cols: list[str], font_size: int, col_gap: int) -> tuple[int, int]:
    """Text block (w, h) of vertical columns."""
    if not cols:
        return 0, 0
    w = len(cols) * font_size + (len(cols) - 1) * col_gap
    h = max(len(c) for c in cols) * round(font_size * 1.05)
    return w, h


def cell_centers(
    cols: list[str], center: tuple[float, float], font_size: int, col_gap: int
) -> list[tuple[str, float, float]]:
    """``(glyph, x, y)`` cell centres; columns run right → left, glyphs top → bottom."""
    w, h = block_size(cols, font_size, col_gap)
    cx, cy = center
    step = round(font_size * 1.05)
    out = []
    for k, col in enumerate(cols):
        x = cx + w / 2 - font_size / 2 - k * (font_size + col_gap)
        for j, ch in enumerate(col):
            out.append((ch, x, cy - h / 2 + step * j + step / 2))
    return out


def draw(draw_ctx, cols: list[str], center, font, col_gap: int, fill="black", **kwargs) -> None:
    for ch, x, y in cell_centers(cols, center, font.size, col_gap):
        draw_ctx.text((x, y), ch, font=font, fill=fill, anchor="mm", **kwargs)
